### ISSUE TITLE
Implicit time integration

### ISSUE THREAD

**Issue by stfnp** [bow-simulation/virtualbow#211]
# Motivation

The time integration method currently used for dynamic analysis is the explicit central difference method. It was initially chosen because it is well known, easy to implement and seemed to work reasonably well.

Although implicit methods haven't yet been tested, it is likely that they would perform better for this type of problem based on the tradeoffs of explicit vs implicit and the usual selection criteria given in literature. The [ADINA Theory and Modelling Guide, Volume I (December 2012)](http://www.adina.com/adinadownloads/docs/tmg-a_89.pdf), for example, states the following:

> 7.5  Choosing between implicit and explicit formulations
> 
> • The main criterion governing the selection of the implicit or explicit formulations is the time scale of the solution.
>  
> • The implicit method can use much larger time steps since it is unconditionally stable. However, it involves the assembly and 
> solution of a system of equations, and it is iterative. Therefore, the computational time per load step is relatively high. The explicit method uses much smaller time steps since it is conditionally stable, meaning that the time step for the solution has to be less than a certain critical time step, which depends on the smallest element size and the material properties. However, it involves no matrix solution and is non-iterative. Therefore, the computational time per load step is relatively low.
> 
> [...]
> 
> • For low-speed dynamic problems, the solution time spans a period of time considerably longer than the time it takes the wave to propagate through an element. The solution in this case is dominated by the lower frequencies of the structure. This class of problems covers most structural dynamics problems, certain metal forming problems, crush analysis, earthquake response and biomedical problems. When the explicit method is used for such problems the resulting number of time steps will be excessive, unless mass-scaling is applied, or the loads are artificially applied over a shorter time frame. No such modifications are needed in the implicit method. Hence, the implicit method is the optimal choice. 
> 
> • For high-speed dynamic problems, the solution time is comparable to the time required for the wave to propagate through the structure. This class of problems covers most wave propagation problems, explosives problems, and high-speed impact problems. For these problems, the number of steps required with the explicit method is not excessive. If the implicit method uses a similar time step it will be much slower and if it uses a much larger time step it will introduce other solution errors since it will not be capturing the pertinent features of the solution (but it will remain stable). Hence, the explicit method is the optimal choice.

Additionally, the following points about the bow model (including planned changes) favor implicit methods
* Damping reduces the maximum stable time step of the central difference method. This is already a problem when (unrealistically) high damping values are chosen.
* Large range of stiffness/frequency in the system ([Wikipedia: Stiff equation](https://en.wikipedia.org/wiki/Stiff_equation))
    * Longitudinal (high frequency) vs bending motion (low frequency) of the limb
    * High frequencies from penalty coefficients
    * Non-uniform sizing of limb elements (#212)

* Switch to the absolute nodal coordinate formulation (#189)
    * Higher order/more accurate elements -> evaluation (possibly) more expensive
    * Consistent (non-diagonal) mass matrix, although still constant

# Requirements

**Adaptive time stepping**

It's not yet clear if this is a strict requirement or a "nice to have". Does the optimal time step actually vary much over the course of the simulation? If so, an adaptive step size could make the simulation more efficient. Another advantage would be that it is easier to find good numerical default settings that work for most users (e.g. a sensible error bound instead of a fixed time step).

Another alternative would be to estimate the timestep once before the simulation, as it is done now.

**Form of the damping terms**

This is not yet clear either. Most method descriptions assume a constant damping matrix, even for otherwise nonlinear problems. A decision needs to be made how damping is handled in the future before a method can be implemented.

* Fully nonlinear damping terms (possible with the current method)
* Rayleigh damping with the initial tangent stiffness matrix
* Rayleigh damping with the current tangent stiffness matrix
* ...

# Literature Review

**[K.J. Bathe: Finite Element Procedures](https://www.amazon.com/dp/0979004950)**

* Houbolt Method
* Wilson-θ Method
    * Special case: Linear Acceleration Method
* Newmark Method
    * Special case: Linear Acceleration Method
    * Special case: Trapezoidal Rule

**[M.A. Chrisfield: Non-linear Finite Element Analysis of Solids and Structures, Volume 2](https://www.wiley.com/en-gb/Non+Linear+Finite+Element+Analysis+of+Solids+and+Structures,+Volume+2,+Advanced+Topics-p-9780470860212)**

* Newmark Method
* Trapezoidal Rule
* Hilber-Hughes-Taylor α Method

**[P. Wriggers: Nichtlineare Finite-Element-Methoden](https://www.springer.com/de/book/9783540677475)**

* Newmark Method
* Bossak Method
* Hilber-Hughes-Taylor α Method

**Interesting links**

https://ckadapa.wordpress.com/2017/06/23/newmark-beta-method-disadvantages/

https://ethz.ch/content/dam/ethz/special-interest/baug/ibk/structural-mechanics-dam/education/femII/presentation_05_dynamics_v3.pdf

https://people.duke.edu/~hpgavin/StructuralDynamics/NumericalIntegration.pdf

Generalized alpha method: https://www.youtube.com/watch?v=wUsTSm-DY1g

# Software Review

Overview of implicit integration methods used in major FEM software packages

| Software     | Methods | Source |
|--------------|---------------|----|
| ANSYS | Newmark<br> Generalized HHT-α | [ANSYS Mechanical APDL Theory Reference (Release 15.0, November 2013)](http://lmee.univ-evry.fr/~hello/ENS/CS93/ANSYS/ANSYS%20Mechanical%20APDL%20Theory%20Reference.pdf) |
| Abaqus | HHT-α<br> Euler backwards | [Abaqus Theory Guide (2016)](http://130.149.89.49:2080/v2016/books/stm/default.htm?startat=ch01s01ath01.html)
| ADINA | Newmark<br> Wilson-θ<br> Bathe | [ADINA Theory and Modelling Guide, Volume I (December 2012)](http://www.adina.com/adinadownloads/docs/tmg-a_89.pdf) |
| Code Aster | Newmark<br> HHT-α<br> Krenk<br> Wilson-θ | [Non-linear transient dynamics analysis](https://www.code-aster.org/V2/UPLOAD/DOC/Formations/05-nonlinear-dynamics.pdf) |
| NX Nastran | Newmark | [NX Nastran Theoretical Manual](https://docs.plm.automation.siemens.com/data_services/resources/nxnastran/10/help/en_US/custom/theoretical/theoretical_NXN.pdf) |
| Msc Nastran | Newmark | [MD/MSC Nastran 2010 Dynamic Analysis User’s Guide](https://simcompanion.mscsoftware.com/resources/sites/MSC/content/meta/DOCUMENTATION/9000/DOC9468/~secure/basdyn.pdf?token=d2DXlc8gIyMgT!5nbI69BHJyhC4NNRRHc9ueuLDAUAGW8XuArWrodjEFEOvfm5PtHasmJy7!-48cHSZTNMV1K7JfwGFKCwsFKoGGsaWVcLE3onbEfG1gtwDBoFc6ajSHwKI-GN55C0Lfp6JV-HYqsaE-rFNydOdH-I3iV1KZlDUUZo2aPYD6S3xGY01LPOsN) |
| Msc Marc | Newmark<br> Houbolt<br> Single Step Houboldt<br> Generalized Alpha | [Marc 2008 r1 Volume A: Theory and User Information](https://simcompanion.mscsoftware.com/resources/sites/MSC/content/meta/DOCUMENTATION/9000/DOC9245/~secure/vola.pdf?token=GE-q2eeF3Pq!TtJ1kHuWJE-CQuo4DmkXXPdfO4q7WCOC1wYpmvR4W0w24wKG0r0USYZEGjjrLCTH9wQfWNJs2TmL9-ukGaumeLK-A-hdDzCrsFzBHRTBF-UrMQWaZIbr4Sm!vgCz022vfMVMEzz9uJh0O-JvtMEW3AlBpDnlhjFUNWXrLOJsb39xEcjRgs9K) |
